src/identifier.rs

```diff
@@ -70,18 +70,83 @@ use crate::alloc::alloc::{alloc, dealloc, Layout};
 #[cfg(no_from_ne_bytes)]
 use crate::backport::FromNeBytes;
 use core::mem;
-use core::num::{NonZeroU64, NonZeroUsize};
+use core::num::NonZeroUsize;
 use core::ptr;
+use core::ptr::NonNull;
 use core::slice;
 use core::str;
 
 #[repr(transparent)]
 pub(crate) struct Identifier {
-    repr: NonZeroU64,
+    repr: Repr,
+}
+
+use crate::identifier::repr::Repr;
+
+mod repr {
+    use core::ptr::NonNull;
+
+    #[cfg(target_pointer_width = "64")]
+    #[derive(PartialEq, Eq, Clone, Copy)]
+    pub struct Repr {
+        ptr: NonNull<u8>,
+    }
+
+    #[cfg(target_pointer_width = "32")]
+    #[derive(PartialEq, Eq, Clone, Copy)]
+    pub struct Repr {
+        ptr: NonNull<u8>,
+        remainder: usize,
```

Comment:
This is going to result in the front and back 4 bytes being randomly swapped depending on what order these two fields get placed in the struct layout.

```rust
fn main() {
    let prerelease = semver::Prerelease::new("asdfjkl").unwrap();
    println!("{}", prerelease);
    println!("{:?}", prerelease.as_str());
}
```

```console
jklasdf
"jkl\0asdf"
```

Reply:
Could we add this one to the list?

![Screenshot from 2022-05-06 10-10-57](https://user-images.githubusercontent.com/1940490/167180074-e0fbedde-a5d1-4cfa-b445-f8b67706172f.png)